src/main/java/com/intellij/rt/debugger/agent/ConditionalBreakpointTransformer.java

```diff
@@ -0,0 +1,367 @@
+            if (methods == null || methods.isEmpty()) {
+                return null;
+            }
+
+            try {
+                final ClassNode classNode = new ClassNode();
+                new ClassReader(classfileBuffer).accept(classNode, 0);
+
+                ClassTransformer transformer = new ClassTransformer(
+                        className, classfileBuffer, ClassWriter.COMPUTE_FRAMES | ClassWriter.COMPUTE_MAXS, loader
+                );
+
+                return transformer.accept(new ClassVisitor(Opcodes.API_VERSION, transformer.writer) {
+                    @Override
+                    public MethodVisitor visitMethod(int access, String name, String descriptor, String signature, String[] exceptions) {
+                        MethodVisitor superMethodVisitor = super.visitMethod(access, name, descriptor, signature, exceptions);
+                        final Map<Integer, InstrumentationBreakpointInfo> lineNumbers = getLineNumbers(name, methods);
+                        if (lineNumbers == null || lineNumbers.isEmpty()) {
+                            return superMethodVisitor;
+                        }
+
+                        MethodNode m = null;
+                        for (MethodNode method : classNode.methods) {
+                            if (method.name.equals(name) && method.desc.equals(descriptor)) {
+                                m = method;
+                            }
+                        }
+                        if (m == null) {
+                            return superMethodVisitor;
+                        }
+
+                        final Map<Integer, InstrumentationBreakpointMappingInfo> argumentMapping = collectArgumentMapping(m, lineNumbers);
+
+                        if (argumentMapping.isEmpty()) {
+                            return superMethodVisitor;
+                        }
+
+                        return new MethodVisitor(api, superMethodVisitor) {
+
+                            @Override
+                            public void visitLineNumber(int line, Label start) {
+                                InstrumentationBreakpointMappingInfo instrumentationBreakpointInfo = argumentMapping.get(line);
+                                if (instrumentationBreakpointInfo != null) {
+                                    addInstrumentationCondition(instrumentationBreakpointInfo);
+                                }
+                                super.visitLineNumber(line, start);
+                            }
+
+                            private void addInstrumentationCondition(InstrumentationBreakpointMappingInfo argumentMapping) {
+                                String fragmentClassName = argumentMapping.inputInfo.fragmentClassName;
+                                int instrumentationId = extractIdFromFragmentClassName(fragmentClassName);
+                                try {
+                                    Label startTry = new Label();
```

Comment:
There should be a link to some knowledge base article or to some IntelliJ class with a JavaDoc that describes the architecture of these calls.